### PR TITLE
add TLS key flags to CLI defaults

### PR DESCRIPTION
**Pull request by anthonybishopric**
This addresses an internal issue related to calling to Consul from a CLI.

### CODE REVIEW

File: pkg/util/net/client.go [hunk 1]
```diff
@@ -31,7 +31,9 @@ func GetTransport(extras map[string]string, rt http.RoundTripper) *headerTranspo
 // on every request.
 func NewHeaderClient(extras map[string]string, rt http.RoundTripper) *http.Client {
 	if len(extras) == 0 {
-		return http.DefaultClient
+		return &http.Client{
```
Comment (bwester) [inline pkg/util/net/client.go]: This seems like an unrelated bug fix. Can you put it in its own commit and discuss the triggers/impact of the bug?

File: pkg/kp/flags/kingpin.go [hunk 2]
```diff
@@ -5,10 +5,11 @@ package flags
 import (
 	"io/ioutil"
 	"log"
+	"net"
 	"net/http"
 
 	"github.com/square/p2/pkg/kp"
-	"github.com/square/p2/pkg/util/net"
+	netutil "github.com/square/p2/pkg/util/net"
```
Comment (bwester) [inline pkg/kp/flags/kingpin.go]: I could swear a different alias for this package is in use elsewhere. Could you look for that alias and make this consistent?

File: pkg/kp/flags/kingpin.go [hunk 3]
```diff
@@ -19,6 +20,9 @@ func ParseWithConsulOptions() (string, kp.Options) {
 	headers := kingpin.Flag("header", "An HTTP header to add to requests, in KEY=VALUE form. Can be specified multiple times.").StringMap()
 	https := kingpin.Flag("https", "Use HTTPS").Bool()
 	wait := kingpin.Flag("wait", "Maximum duration for Consul watches, before resetting and starting again.").Default("30s").Duration()
+	caFile := kingpin.Flag("tls-ca-file", "File containing the x509 PEM-encoded CA ").ExistingFile()
+	keyFile := kingpin.Flag("tls-key-file", "File containing the x509 PEM-encoded private key").ExistingFile()
```
Comment (bwester) [inline pkg/kp/flags/kingpin.go]: Please verify that a private key file is x509 PEM-encoded.

File: pkg/kp/flags/kingpin.go [hunk 4]
```diff
@@ -29,10 +33,31 @@ func ParseWithConsulOptions() (string, kp.Options) {
 		}
 		*token = string(tokenBytes)
 	}
+	var transport http.RoundTripper
+	if *caFile != "" || *keyFile != "" || *certFile != "" {
```
Comment (bwester) [inline pkg/kp/flags/kingpin.go]: I'm not sure why you want all or nothing. It seems reasonable to me that one might use the default certificate authorities while supplying a client certificate.

File: pkg/kp/flags/kingpin.go [hunk 5]
```diff
@@ -29,10 +33,31 @@ func ParseWithConsulOptions() (string, kp.Options) {
...
+		if *caFile == "" || *keyFile == "" || *certFile == "" {
```
Comment (bwester) [inline pkg/kp/flags/kingpin.go]: GetTLSConfig already validates its arguments; this check seems redundant.

File: bin/p2-wipe-reality/main.go [hunk 6]
```diff
@@ -19,6 +21,16 @@ a known state.
 `
 )
 
+func handleError(err error, message string) {
```
Comment (bwester) [inline bin/p2-wipe-reality/main.go]: I can see the benefit in wanting to unpack the hidden error messages. But:

1. What if I eventually want to do something other than print the error? Can you unlock without side effects?
2. This is unrelated to adding TLS flags. Can you put it in its own commit?

File: pkg/kp/flags/kingpin.go [hunk 7]
```diff
@@ -19,6 +20,9 @@ func ParseWithConsulOptions() (string, kp.Options) {
 	headers := kingpin.Flag("header", "An HTTP header to add to requests, in KEY=VALUE form. Can be specified multiple times.").StringMap()
 	https := kingpin.Flag("https", "Use HTTPS").Bool()
 	wait := kingpin.Flag("wait", "Maximum duration for Consul watches, before resetting and starting again.").Default("30s").Duration()
+	caFile := kingpin.Flag("tls-ca-file", "File containing the x509 PEM-encoded CA ").ExistingFile()
```
Comment (bwester) [inline pkg/kp/flags/kingpin.go]: Having "tls-" seems wordy, and I don't want to ever type that.  What do you think about removing that prefix? Incidentally, would that make these flags the same as for curl?

File: bin/p2-wipe-reality/main.go (hunk 6 above)
Comment (anthonybishopric) [inline bin/p2-wipe-reality/main.go]: I can remove it, it was meant for debugging why this utility wasn't working. Dealing with the unsafe errors in Consul isn't something I care to solve in this PR.

File: pkg/kp/flags/kingpin.go (hunk 2 above)
Comment (anthonybishopric) [inline pkg/kp/flags/kingpin.go]: It's mapped to `p2net` in a different repo at Square. Still want it to be consistent?

File: pkg/kp/flags/kingpin.go (hunk 7 above)
Comment (anthonybishopric) [inline pkg/kp/flags/kingpin.go]: The curl options are `--cert`, `--cacert` and `--key`. For at least the last option, `key` is a fairly ambiguous option given that it could refer to a key in the kv store or a gpg signing key. I can remove the `tls-` prefix and I think that will make it sufficiently less verbose.

File: pkg/kp/flags/kingpin.go (hunk 3 above)
Comment (anthonybishopric) [inline pkg/kp/flags/kingpin.go]: Will the `util/net` helper not do that for me?

File: pkg/util/net/client.go (hunk 1 above)
Comment (anthonybishopric) [inline pkg/util/net/client.go]: extracted

File: pkg/kp/flags/kingpin.go (hunk 3 above)
Comment (bwester) [inline pkg/kp/flags/kingpin.go]: Sorry, not sure what I was thinking. What you have will work just fine.

File: pkg/kp/flags/kingpin.go (hunk 7 above)
Comment (bwester) [inline pkg/kp/flags/kingpin.go]: You're right, there are a lot of other keys in the system. Sure, keep `tls-`.

File: pkg/kp/flags/kingpin.go (hunk 2 above)
Comment (bwester) [inline pkg/kp/flags/kingpin.go]: Perhaps the real solution is to not have our own package named "net".

Pick your favorite.

**Comment by bwester** [comment]
:ship:
